core/domain/topic_services.py

```diff
@@ -215,16 +216,20 @@ def get_topics_by_ids(topic_ids):
     return topics
 
 
-def get_topic_by_name(topic_name):
+def get_topic_by_name(topic_name, quoted=False):
     """Returns a domain object representing a topic.
 
     Args:
         topic_name: str. The name of the topic.
+        quoted: boolean. True if the topic_name is url-encoded. False otherwise.
 
     Returns:
         Topic or None. The domain object representing a topic with the
         given id, or None if it does not exist.
     """
+    if quoted:
+        topic_name = urllib.unquote_plus(topic_name)
```

Comment:
This seems off from an architecture perspective -- url parsing should happen only at the controlller layer (and throughout the rest of the app, it should be the "true value"). Could we keep all the transformation operations at the controller level?

Reply:
Ah, but this goes against how we conceive *_services.py files typically throughout the app, though. If you look at the [codebase overview](https://github.com/oppia/oppia/wiki/Overview-of-the-Oppia-codebase), the *_services.py and *_domain.py live in the "logic/domain layer", where everything has been fully transformed/sanitized/etc. and we are working exclusively with the real underlying values. This practice helps shield the logic + storage layers from the normalization/parsing/validation which gets done at the controller level (and which is also a place where bad inputs get rejected). Note that url quoting/unquoting is exclusively needed only at that boundary, since it's how the JS code and the backend communicate; it's not needed further down (and shouldn't be).

So, if we want a function to handle the URL decoding, that's fine, but let's put it in the controller layer as a method in the relevant handler.